server/tools.go

```diff
@@ -0,0 +1,414 @@
+		p.ParserState = ToolCallSendTokens
+	case p.state == NotPartialPrefix:
+		p.ParserState = ToolCallSendPartial
+	default:
+		p.ParserState = ToolCallSendTokens
+		p.sb.Reset()
+		p.state = SendTokens
+	}
+}
+
+// string, and if it has a prefix
+func (p *ToolParser) checkPrefix(s string) (string, bool) {
+	fmt.Printf("checking prefix: input=%s prefix=%s\n", s, p.toolPrefix)
+
+	if p.toolPrefix == "" {
+		return s, true
+	}
+	original := s
+	s, hasPrefix := strings.CutPrefix(s, p.toolPrefix)
+	if hasPrefix {
+		p.state = ForceTools
+		fmt.Printf("found exact prefix match: remaining=%s\n", s)
+		// partial tool possibly - accumulate
+	} else if suffixOverlap(s, p.toolPrefix) > 0 {
+		p.state = PartialPrefix
+		fmt.Printf("found partial prefix: remaining=%s\n", s)
+		return "", false
+		// the case where "token<tool_call>" - send "token" back
+		// accounts for spaces in prefix or suffix to avoid breaking cache
+	} else if strings.Contains(original, p.toolPrefix) {
+		idx := strings.Index(original, p.toolPrefix)
+		if idx != -1 {
+			// still keeps the prefix
+			p.state = ContainsPrefix
+			p.sb.Reset()
+			// todo: see if there is a simpler way for this
+			idx2 := strings.Index(s, p.toolPrefix)
+			// buffer now only has the prefix
+			p.sb.WriteString(s[idx2:])
+			fmt.Printf("found prefix in middle: prefix_start=%d content_before=%s\n", idx, original[:idx])
+			return original[:idx], false
+		}
+	}
+
+	return s, true
+}
+
+// TODO: simplify the flow of this function
+// ParseToolCalls extracts tool calls from a string using a tool token prefix or direct JSON parsing.
+// Returns tool calls, whether parsing is incomplete, and any errors.
+func (p *ToolParser) ParseToolCalls(s string) ([]api.ToolCall, string) {
+	fmt.Printf("parsing tool calls: input=%s current_state=%s\n", s, p.state)
+
+	p.sb.WriteString(s)
+	s = p.sb.String()
+
+	s = strings.TrimSpace(s)
+
+	if len(s) == 0 {
+		p.updateExternalState(nil)
+		return nil, ""
+	}
+
+	s, cont := p.checkPrefix(s)
+	if !cont {
+		p.updateExternalState(nil)
+		if p.state == ContainsPrefix {
+			fmt.Printf("returning partial prefix: remaining=%s\n", s)
+			return nil, s
+		}
+		// * we'd be returning here for just accumulating with possible prefix
+		// * ext state is accumulation
+		return nil, ""
+	}
+	// * lets say the check fails here and now we're still in external state accumulation here
+
+	// stay in SendTokens unless we have a prefix
+	if p.state == SendTokens {
+		p.updateExternalState(nil)
+		p.sb.Reset()
+		fmt.Printf("returning send tokens: remaining=%s\n", s)
+		return nil, s
+	}
+
+	// * we'd parse here as json to see if it's a tool call
+	tcs, partial, ok := p.parseJSONToolCalls(s)
+	// * it would not be a tool call here
+	p.updateStateAfterJSONParse(ok, partial, tcs)
+	if !ok {
+		// * and so we should send the data here
+		// * we also need to move out of that internal state after sending the tokens
+		if p.state == NotPartialPrefix {
+			p.state = SendTokens
```

Comment:
going to try to avoid state management in this func and try to keep it to just the control flow